### PR TITLE
Enable specifying  master_authorized_networks_config

### LINKED ISSUE #7: Support for Master Auth
They seem to not be supported in this module... will they be in the future?

### PR DESCRIPTION
**Pull request by pratikmallya**
Fixes #7

### CODE REVIEW

**Comment by ryanckoch** [comment]
The `make -s` tests currently fail when the the `test/integration/tmp` directory has already been generated. The tests that matter most in  this instance are the `make test_integration` tests. I'll submit a separate PR for better clarification in the README.

**Comment by pratikmallya** [comment]
@ryanckoch there was a typo which has been fixed. Could you take a look?

File: outputs.tf [hunk 1]
```diff
@@ -50,6 +50,11 @@ output "min_master_version" {
   value       = "${local.cluster_min_master_version}"
 }
 
+output "master_authorized_networks_config" {
+  description = "Networks from which access to master is permitted"
+  value       = "${local.master_authorized_networks_config}"
```
[the remark this answers is not in the document]
Comment (pratikmallya) [reply outputs.tf]: Ah, thanks @morgante ! Fixed

**Comment by ryanckoch** [comment]
Looking good, just currently hitting:
```
       Error: Error running plan: 1 error(s) occurred:

       * module.gke.local.cluster_type_output_master_authorized_networks_config: local.cluster_type_output_master_authorized_networks_config: Resource 'google_container_cluster.primary' does not have attribute 'master_authorized_networks_config' for variable 'google_container_cluster.primary.*.master_authorized_networks_config'
```
This is a similar issue to what we've seen in on another module: https://github.com/terraform-google-modules/terraform-google-network/issues/5

This seems to be caused by the provider ignoring the empty `master_authorized_networks_config = "${var.master_authorized_networks_config}"` in the `google_container_cluster` resource (as it should). It in turn doesn't create that attribute on the resource in the terraform state file.

The best way we've come up to get around this in this case, is just pass the variable to the output. Would also in this case negate the need to set the `cluster_type_output_master_authorized_networks_config` local variable.

**Comment by pratikmallya** [comment]
Updated

**Comment by morgante** [comment]
Could we add a test for this to the integration tests?

**Comment by ryanckoch** [comment]
Tests are all passing now with `master_authorized_networks_config` disabled.

**Comment by ryanckoch** [comment]
Do `cp test/integration/gcloud/sample.sh test/integration/gcloud/config.sh`
and modify the first block of variables to fit your environment.

**Comment by morgante** [comment]
@ryanckoch Can we add a sample `config.sh` and integration test setup directions?

**Comment by ryanckoch** [comment]
There is a sample, it just needs documented a bit better.
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/12

**Comment by qvallance** [comment]
I tackled this a little while ago and found the formatting required for the variable to be a bit tricky (at least, how I got it to work), so I ended up including a full example in my variable description to help users.
```
variable "master_authorized_networks_config" {
  type        = "list"
  description = <<EOF
  The desired configuration options for master authorized networks. Omit the nested cidr_blocks attribute to disallow external access (except the cluster node IPs, which GKE automatically whitelists)

  ### example format ###
  default = [{
    cidr_blocks = [{
      cidr_block   = "10.0.0.0/8"
      display_name = "example_network"
    }],
  }]

  EOF
  default = []
}
```

**Comment by pratikmallya** [comment]
@qvallance I like the idea of being more explicit; this is kinda of a downstream module though (the variable is a pass-through input to the [google_container_cluster module](https://www.terraform.io/docs/providers/google/r/container_cluster.html#master_authorized_networks_config)) so I'm not sure if we want to be that descriptive.

**Comment by qvallance** [comment]
@pratikmallya I generally agree that downstream variables don’t need to be that descriptive.  I just found the required pass through format for this particularly variable not obvious from the documentation.

**Comment by pratikmallya** [comment]
Ah, I see what you're saying. It doesn't match what the original variable looks like.... thanks @qvallance !

**Comment by pratikmallya** [comment]
Need more help trying to run the tests locally. Stuck on this issue with the terraform driver now:
```
$ make test_integration
./test/integration/gcloud/run.sh regional

Creating main.tf file
Creating outputs.tf file
Using public_suffix 3.0.3
Using addressable 2.5.2
Using aws-sigv4 1.0.3
Using jmespath 1.4.0
Using aws-sdk-core 2.11.134
Using aws-sdk-resources 2.11.134
Using aws-sdk 2.11.134
Using concurrent-ruby 1.0.5
Using multipart-post 2.0.0
Using faraday 0.15.3
Using unf_ext 0.0.7.5
Using unf 0.1.4
Using domain_name 0.5.20180417
Using http-cookie 1.0.3
Using faraday-cookie_jar 0.0.6
Using timeliness 0.3.8
Using ms_rest 0.7.2
Using ms_rest_azure 0.11.0
Using azure_graph_rbac 0.17.0
Using azure_mgmt_resources 0.17.1
Using builder 3.2.3
Using bundler 1.16.0
Using coderay 1.1.2
Using declarative 0.0.10
Using declarative-option 0.1.0
Using diff-lcs 1.3
Using excon 0.62.0
Using multi_json 1.13.1
Using docker-api 1.34.2
Using dry-configurable 0.7.0
Using dry-container 0.6.0
Using dry-core 0.4.7
Using dry-equalizer 0.2.1
Using dry-inflector 0.1.2
Using dry-logic 0.4.2
Using dry-types 0.13.2
Using dry-validation 0.12.2
Using erubis 2.7.0
Using faraday_middleware 0.12.2
Using ffi 1.9.25
Using jwt 2.1.0
Using memoist 0.16.0
Using os 1.0.0
Using signet 0.9.2
Using googleauth 0.6.6
Using httpclient 2.8.3
Using mime-types-data 3.2018.0812
Using mime-types 3.2.2
Using uber 0.1.0
Using representable 3.0.4
Using retriable 3.1.2
Using google-api-client 0.19.8
Using gssapi 1.2.0
Using gyoku 1.3.1
Using hashie 3.6.0
Using htmlentities 4.3.4
Using inifile 3.0.0
Using json 2.1.0
Using method_source 0.9.0
Using mixlib-log 2.0.4
Using parallel 1.12.1
Using parslet 1.8.2
Using pry 0.11.3
Using rspec-support 3.8.0
Using rspec-core 3.8.0
Using rspec-expectations 3.8.1
Using rspec-mocks 3.8.0
Using rspec 3.8.0
Using rspec-its 1.2.0
Using rubyzip 1.2.2
Using semverse 2.0.0
Using sslshake 1.2.0
Using thor 0.20.0
Using tomlrb 1.2.7
Using mixlib-shellout 2.4.0
Using net-ssh 4.2.0
Using net-scp 1.2.1
Using little-plugger 1.1.4
Using logging 2.2.2
Using nori 2.6.0
Using rubyntlm 0.6.2
Using winrm 2.2.3
Using winrm-fs 1.3.0
Using train 1.4.37
Using inspec 2.2.112
Using mixlib-versioning 1.2.2
Using mixlib-install 3.11.5
Using net-ssh-gateway 1.3.0
Using winrm-elevated 1.1.0
Using test-kitchen 1.23.2
Using kitchen-inspec 0.23.1 from https://github.com/inspec/kitchen-inspec.git (at 0590f1b@0590f1b)
Using kitchen-terraform 3.3.1
Bundle complete! 3 Gemfile dependencies, 92 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Ignoring ffi-1.9.25 because its extensions are not built.  Try: gem pristine ffi --version 1.9.25
Ignoring ffi-yajl-2.3.1 because its extensions are not built.  Try: gem pristine ffi-yajl --version 2.3.1
Ignoring nokogiri-1.6.8 because its extensions are not built.  Try: gem pristine nokogiri --version 1.6.8
Ignoring yajl-ruby-1.4.0 because its extensions are not built.  Try: gem pristine yajl-ruby --version 1.4.0
-----> Starting Kitchen (v1.13.2)
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ClientError
>>>>>> Message: Could not load the 'terraform' driver from the load path. Please ensure that your driver is installed as a gem or included in your Gemfile if using Bundler.
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration

make: *** [regional_test_integration] Error 20
```
Even though I do have terraform installed. Ideas on how to fix this?

**Comment by pratikmallya** [comment]
@morgante without giving clear instructions on how to run the integration tests, its hard for me to make the changes that you have requested. I'm unfortunately not very familiar with ruby/chef/kitchen and have not been able to figure how to fix this error

File: variables.tf [hunk 2]
```diff
@@ -65,6 +65,12 @@ variable "node_version" {
   default     = ""
 }
 
+variable "master_authorized_networks_config" {
+  description = "The desired configuration options for master authorized networks. Omit the nested cidr_blocks attribute to disallow external access (except the cluster node IPs, which GKE automatically whitelists)."
```
Comment (morgante) [inline variables.tf]: @pratikmallya Could we expand this to include an example config, like here: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/10#issuecomment-423255873

**Comment by morgante** [comment]
@pratikmallya Understood, I'm okay with merging this without tests assuming our existing tests pass.

If you address this final comment we can merge: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/10#pullrequestreview-157822705

Thanks for your work on this and patience.

**Comment by pratikmallya** [comment]
@morgante @ryanckoch updated the variable description

**Comment by pratikmallya** [comment]
@morgante ping

**Pull request merged**